File: Habibi/src/api/kb.ts
// Knowledge Base (RAG) — data access seam.
// KB-1: retrieve. KB-2: documents library + stats + upload/reindex.

import { useQuery } from "@tanstack/react-query";

import type { FaqPair, KbChunk, KbDocType, KbDocument, RetrievalResult } from "@/api/types/kb";
import { apiDelete, apiGet, apiPatch, apiPost, apiUpload } from "./config";

export type { FaqPair, KbChunk, KbDocument, RetrievalResult };

export interface KbGap {
  id: string;
  text: string;
  hits: number;
  lastSeen: string;
  topIntent: string;
  hasKbDoc: boolean;
  hasFaq: boolean;
  resolved: boolean;
  suggestedFix: "kb" | "prompt" | "both";
  linkedDocumentId?: string | null;
  linkedFaqId?: string | null;
  linkedPromptVersionId?: string | null;
}

export interface KbRetrieveRequest {
  query: string;
  topK?: number;
  includeDraftAnswer?: boolean;
  source?: string;
}

export interface KbRetrieveResponse {
  results: RetrievalResult[];
  draftAnswer: string | null;
  latencyMs: number;
  embeddingModel: string;
  chatModel: string | null;
  logId: string;
}

export interface KbStats {
  docs: number;
  activeDocs: number;
  faqs: number;
  chunks: number;
  gaps: number;
  lastIndexed: string;
  avgScore: number;
}

export interface KbUploadResult {
  document: KbDocument;
  jobId: string | null;
}

export interface KbReindexResult {
  jobId: string;
  documentId: string;
  status: string;
}

export interface KbIndexJob {
  id: string;
  documentId: string;
  status: string;
  chunkSize: number | null;
  chunkOverlap: number | null;
  embeddingModel: string | null;
  startedAt: string | null;
  completedAt: string | null;
  error: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface KbUploadInput {
  file: File;
  title?: string;
  type: KbDocType;
  chunkSize: number;
  overlap: number;
  indexNow: boolean;
  tags?: string[];
}

export async function retrieveKb(input: KbRetrieveRequest): Promise<KbRetrieveResponse> {
  const topK = input.topK ?? 4;
  const includeDraftAnswer = input.includeDraftAnswer ?? true;

  return apiPost<KbRetrieveResponse>("/kb/retrieve", {
    query: input.query,
    topK,
    includeDraftAnswer,
    source: input.source ?? "test",
  });
}

export async function fetchKbStats(): Promise<KbStats> {
  return apiGet<KbStats>("/kb/stats");
}

export function useKbStats() {
  return useQuery({ queryKey: ["kb", "stats"], queryFn: fetchKbStats });
}

export async function fetchKbDocuments(): Promise<KbDocument[]> {
  return apiGet<KbDocument[]>("/kb/documents");
}

export function useKbDocuments() {
  return useQuery({ queryKey: ["kb", "documents"], queryFn: fetchKbDocuments, staleTime: 10_000 });
}

export async function fetchKbChunks(documentId: string): Promise<KbChunk[]> {
  return apiGet<KbChunk[]>(`/kb/documents/${documentId}/chunks`);
}

export function useKbChunks(documentId: string | null) {
  return useQuery({
    queryKey: ["kb", "chunks", documentId],
    queryFn: () => fetchKbChunks(documentId!),
    enabled: Boolean(documentId),
    staleTime: 10_000,
  });
}

export async function patchKbDocument(
  id: string,
  patch: {
    enabled?: boolean;
    title?: string;
    tags?: string[];
    chunkSize?: number;
    overlap?: number;
  },
): Promise<KbUploadResult> {
  return apiPatch<KbUploadResult>(`/kb/documents/${id}`, patch);
}

export async function reindexKbDocument(id: string): Promise<KbReindexResult> {
  return apiPost<KbReindexResult>(`/kb/documents/${id}/reindex`, {});
}

export interface KbSnapshot {
  id: string;
  label: string;
  documentIds: string[];
  faqIds: string[];
  documentCount: number;
  faqCount: number;
  createdAt: string | null;
}

export async function reindexAllKbDocuments(): Promise<{
  jobIds: string[];
  count: number;
  snapshot?: KbSnapshot | null;
}> {
  return apiPost<{ jobIds: string[]; count: number; snapshot?: KbSnapshot | null }>(
    "/kb/reindex-all",
    {},
  );
}

export type KbPurgeScope = "all" | "uploads" | "corpus";

export interface KbDeleteDocumentResult {
  deleted: boolean;
  documentId: string;
  faqsDeleted: number;
  minioObjectsRemoved: number;
}

export interface KbPurgeResult {
  scope: KbPurgeScope;
  documentsDeleted: number;
  faqsDeleted: number;
  minioObjectsRemoved: number;
  documentIds: string[];
}

export interface KbIngestSourceDbResult {
  products: string[];
  jobsDrained: number;
  faqsUpserted: number;
  docs: number;
  chunks: number;
  faqs: number;
}

export async function deleteKbDocument(id: string): Promise<KbDeleteDocumentResult> {
  return apiDelete<KbDeleteDocumentResult>(`/kb/documents/${id}`);
}

export async function purgeKbDocuments(scope: KbPurgeScope): Promise<KbPurgeResult> {
  return apiPost<KbPurgeResult>("/kb/documents/purge", { scope, confirm: true });
}

export async function ingestSourceDb(product?: string): Promise<KbIngestSourceDbResult> {
  const q = product ? `?product=${encodeURIComponent(product)}` : "";
  return apiPost<KbIngestSourceDbResult>(`/kb/ingest/source-db${q}`, {});
}

export async function uploadKbDocument(input: KbUploadInput): Promise<KbUploadResult> {
  const form = new FormData();
  form.append("file", input.file);
  form.append("title", input.title ?? "");
  form.append("type", input.type);
  form.append("chunkSize", String(input.chunkSize));
  form.append("overlap", String(input.overlap));
  form.append("indexNow", String(input.indexNow));
  form.append("tags", JSON.stringify(input.tags ?? []));
  return apiUpload<KbUploadResult>("/kb/documents", form);
}

export async function uploadKbDocumentVersion(id: string, file: File): Promise<KbUploadResult> {
  const form = new FormData();
  form.append("file", file);
  return apiUpload<KbUploadResult>(`/kb/documents/${id}/versions`, form);
}

export async function fetchKbIndexJob(jobId: string): Promise<KbIndexJob> {
  return apiGet<KbIndexJob>(`/kb/index-jobs/${jobId}`);
}

/** Poll until job leaves queued/running (or timeout). */
export async function pollKbIndexJob(
  jobId: string,
  opts: { intervalMs?: number; timeoutMs?: number } = {},
): Promise<KbIndexJob> {
  const intervalMs = opts.intervalMs ?? 1500;
  const timeoutMs = opts.timeoutMs ?? 120_000;
  const started = Date.now();
  while (Date.now() - started < timeoutMs) {
    const job = await fetchKbIndexJob(jobId);
    if (job.status === "succeeded" || job.status === "failed") return job;
    await new Promise((r) => setTimeout(r, intervalMs));
  }
  throw new Error(`index job ${jobId} timed out`);
}

/** Poll many index jobs with bounded concurrency; returns settled results (never throws). */
export async function pollKbIndexJobs(
  jobIds: string[],
  opts: { intervalMs?: number; timeoutMs?: number; concurrency?: number } = {},
): Promise<{ succeeded: number; failed: number; timedOut: number; jobs: KbIndexJob[] }> {
  const concurrency = Math.max(1, Math.min(opts.concurrency ?? 4, jobIds.length || 1));
  const jobs: KbIndexJob[] = new Array(jobIds.length);
  let cursor = 0;

  async function worker() {
    while (cursor < jobIds.length) {
      const i = cursor;
      cursor += 1;
      const id = jobIds[i]!;
      try {
        jobs[i] = await pollKbIndexJob(id, opts);
      } catch {
        jobs[i] = {
          id,
          documentId: "",
          status: "failed",
          chunkSize: null,
          chunkOverlap: null,
          embeddingModel: null,
          startedAt: null,
          completedAt: null,
          error: "timed out or unreachable",
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString(),
        } satisfies KbIndexJob;
      }
    }
  }

  await Promise.all(Array.from({ length: Math.min(concurrency, jobIds.length) }, () => worker()));

  let succeeded = 0;
  let failed = 0;
  let timedOut = 0;
  for (const job of jobs) {
    if (job.status === "succeeded") succeeded += 1;
    else if (job.error?.includes("timed out")) timedOut += 1;
    else failed += 1;
  }
  return { succeeded, failed, timedOut, jobs };
}

/**
 * Approximate token windows from real file text (word≈token) for upload preview.
 * Server indexing still uses tiktoken — this is a local estimate only.
 */
export function previewChunksFromText(
  text: string,
  chunkSize: number,
  overlap: number,
  maxChars = 500_000,
): { count: number; samples: string[] } {
  const bounded = (text || "").slice(0, maxChars);
  const tokens = bounded.trim().split(/\s+/).filter(Boolean);
  if (tokens.length === 0) return { count: 0, samples: [] };
  const size = Math.max(1, chunkSize);
  const ov = Math.max(0, Math.min(overlap, size - 1));
  const step = Math.max(1, size - ov);
  const samples: string[] = [];
  let count = 0;
  for (let start = 0; start < tokens.length; start += step) {
    const slice = tokens.slice(start, start + size);
    if (!slice.length) break;
    count += 1;
    if (samples.length < 5) {
      const joined = slice.join(" ");
      samples.push(joined.length > 220 ? `${joined.slice(0, 217)}…` : joined);
    }
    if (start + size >= tokens.length) break;
  }
  return { count: Math.max(1, count), samples };
}

// --- KB-3: FAQs + Gaps ---

export async function fetchKbFaqs(): Promise<FaqPair[]> {
  return apiGet<FaqPair[]>("/kb/faqs");
}

export function useKbFaqs() {
  return useQuery({ queryKey: ["kb", "faqs"], queryFn: fetchKbFaqs, staleTime: 10_000 });
}

export async function createKbFaq(input: {
  question: string;
  answer: string;
  intent: string;
  enabled?: boolean;
  linkedDocId?: string;
  gapId?: string;
}): Promise<FaqPair> {
  return apiPost<FaqPair>("/kb/faqs", input);
}

export async function patchKbFaq(
  id: string,
  patch: {
    question?: string;
    answer?: string;
    intent?: string;
    enabled?: boolean;
    linkedDocId?: string | null;
  },
): Promise<FaqPair> {
  return apiPatch<FaqPair>(`/kb/faqs/${id}`, patch);
}

export async function deleteKbFaq(id: string): Promise<void> {
  await apiDelete(`/kb/faqs/${id}`);
}

export async function fetchKbGaps(): Promise<KbGap[]> {
  return apiGet<KbGap[]>("/kb/gaps");
}

export function useKbGaps() {
  return useQuery({ queryKey: ["kb", "gaps"], queryFn: fetchKbGaps, staleTime: 10_000 });
}

export async function linkKbGap(
  gapId: string,
  link: { faqPairId?: string; kbDocumentId?: string; promptVersionId?: string },
): Promise<KbGap> {
  return apiPost<KbGap>(`/kb/gaps/${gapId}/link`, link);
}

/** Mock promote returns a synthetic id that is not a real skill row. */

export async function promoteGapToSkill(
  gapId: string,
): Promise<{ id: string; slug: string; signatureStatus?: string }> {
  return apiPost(`/kb/gaps/${gapId}/promote-skill`, {});
}

export async function fetchKbSnapshots(): Promise<KbSnapshot[]> {
  return apiGet<KbSnapshot[]>("/kb/snapshots");
}

export function useKbSnapshots() {
  return useQuery({ queryKey: ["kb", "snapshots"], queryFn: fetchKbSnapshots });
}

File: backend/schemas/kb.py
"""Knowledge base: documents, FAQs, retrieval, snapshots.

One module of the ``schemas`` package (was one 6,400-line file); the
router of the same name serves these. ``schemas/__init__`` re-exports every name.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# The authored flow graph is a domain model, not a transport shape — it is
# shared verbatim by the API, the validator and the voice runtime, so it is
# defined once in flow_graph and reused here rather than restated.
from flow_graph import FlowGraph, FlowIssue, FlowValidation  # noqa: F401

# ---------------------------------------------------------------------------
# Knowledge Base / RAG (Phase KB-1 — retrieve spine)
# ---------------------------------------------------------------------------


class KbRetrievalResultItem(BaseModel):
    """Mirrors Habibi RetrievalResult in kb-seed.ts."""

    model_config = ConfigDict(extra="forbid")

    chunkId: str
    docId: str
    docTitle: str
    docType: str | None = None
    heading: str
    snippet: str
    score: float
    matchedTerms: list[str]


class KbRetrieveRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    query: str
    topK: int = Field(default=4, ge=1, le=20)
    includeDraftAnswer: bool = True
    source: str = "test"


class KbRetrieveResponse(BaseModel):
    """The full shape of ``kb_retrieve.retrieve()``, not a subset of it.

    ``extra="forbid"`` is right and stays. What was wrong is that retrieval grew
    four fields — the top1-top2 margin, the per-stage timing split, whether the
    reranker ran, and whether the row came from the result cache — and this model
    was never told. Every one of them is present on both the fresh and the cached
    path, so ``POST /kb/retrieve`` returned a hard 500 on every call, and the
    Test Retrieval screen an operator would use to diagnose a bad answer was the
    one surface that could not produce one.

    A response model that describes less than the handler returns is not a
    stricter contract, it is a broken endpoint. Keep these in step.
    """

    model_config = ConfigDict(extra="forbid")

    results: list[KbRetrievalResultItem]
    draftAnswer: str | None = None
    latencyMs: int
    embeddingModel: str
    chatModel: str | None = None
    logId: str
    #: Top1-top2 score gap. The reported confidence signal (the absolute top
    #: score predicts retrieval success at AUC 0.548 — a coin flip), and what
    #: the KB-gap screen thresholds on.
    margin: float = 0.0
    #: Per-stage milliseconds: rate_ms, embed_ms, ann_ms, rerank_ms, draft_ms.
    stageMs: dict[str, float] = Field(default_factory=dict)
    reranked: bool = False
    cached: bool = False


KbDocType = Literal["policy", "sop", "product", "compliance", "faq", "benefits"]


KbDocStatus = Literal["draft", "indexing", "indexed", "stale", "failed"]


class KbDocumentResponse(BaseModel):
    """Mirrors Habibi KbDocument in kb-seed.ts."""

    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    filename: str
    type: KbDocType
    version: str
    status: KbDocStatus
    enabled: bool
    chunks: int
    chunkSize: int
    overlap: int
    embeddingModel: str
    updatedBy: str
    lastIndexed: str
    tags: list[str]


class KbChunkResponse(BaseModel):
    """Mirrors Habibi KbChunk."""

    model_config = ConfigDict(extra="forbid")

    id: str
    docId: str
    index: int
    heading: str
    tokens: int
    text: str
    hits: int


class KbStatsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    docs: int
    activeDocs: int
    faqs: int
    chunks: int
    gaps: int
    lastIndexed: str
    avgScore: float


class KbDocumentPatchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool | None = None
    title: str | None = None
    tags: list[str] | None = None
    chunkSize: int | None = Field(default=None, ge=64, le=4096)
    overlap: int | None = Field(default=None, ge=0, le=1024)


class KbReindexResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    jobId: str
    documentId: str
    status: str = "queued"


class KbIndexJobResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    documentId: str
    status: str
    chunkSize: int | None = None
    chunkOverlap: int | None = None
    embeddingModel: str | None = None
    startedAt: str | None = None
    completedAt: str | None = None
    error: str | None = None
    createdAt: str
    updatedAt: str


class KbUploadResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document: KbDocumentResponse
    jobId: str | None = None


class KbFaqResponse(BaseModel):
    """Mirrors Habibi FaqPair in kb-seed.ts."""

    model_config = ConfigDict(extra="forbid")

    id: str
    question: str
    answer: str
    intent: str
    enabled: bool
    updatedAt: str
    linkedDocId: str | None = None


class KbFaqCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    question: str
    answer: str
    intent: str = "other"
    enabled: bool = True
    linkedDocId: str | None = None
    gapId: str | None = None  # optional: link analytics gap on create


class KbFaqPatchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    question: str | None = None
    answer: str | None = None
    intent: str | None = None
    enabled: bool | None = None
    linkedDocId: str | None = None


KbGapSuggestedFix = Literal["kb", "prompt", "both"]


class KbGapResponse(BaseModel):
    """Coverage gap row — mirrors Habibi UnansweredQuestion (+ link state)."""

    model_config = ConfigDict(extra="forbid")

    id: str
    text: str
    hits: int
    lastSeen: str
    topIntent: str
    hasKbDoc: bool
    hasFaq: bool
    resolved: bool
    suggestedFix: KbGapSuggestedFix
    linkedDocumentId: str | None = None
    linkedFaqId: str | None = None
    linkedPromptVersionId: str | None = None


class KbGapLinkRequest(BaseModel):
    """Exactly one of faqPairId | kbDocumentId | promptVersionId."""

    model_config = ConfigDict(extra="forbid")

    faqPairId: str | None = None
    kbDocumentId: str | None = None
    promptVersionId: str | None = None


class KbSnapshotCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    label: str | None = None


class KbSnapshotResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    label: str
    documentIds: list[str] = []
    faqIds: list[str] = []
    documentCount: int = 0
    faqCount: int = 0
    createdAt: str | None = None


KbPurgeScope = Literal["all", "uploads", "corpus"]


class KbPurgeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scope: KbPurgeScope = "uploads"
    confirm: bool = False


class KbPurgeResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scope: KbPurgeScope
    documentsDeleted: int
    faqsDeleted: int
    minioObjectsRemoved: int = 0
    documentIds: list[str] = []


class KbDeleteDocumentResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    deleted: bool
    documentId: str
    faqsDeleted: int = 0
    minioObjectsRemoved: int = 0


class KbIngestSourceDbResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    products: list[str]
    jobsDrained: int
    faqsUpserted: int
    docs: int
    chunks: int
    faqs: int


# ── KB ───────────────────────────────────────────────────────────────────────


class KbReindexAllResponse(BaseModel):
    """`reindex_all_kb_documents` plus the snapshot hook; `snapshot` is None
    when the snapshot could not be taken (the reindex itself succeeded)."""

    jobIds: list[str]
    count: int
    snapshot: KbSnapshotResponse | None = None
